### PR TITLE
Better Python errors using source python line numbers and filenames

### PR DESCRIPTION
**Pull request by mondus**
Pure Python agent functions have always had horrible error reporting for compilation errors. This PR improves upon these by making the compiler (nvcc) report errors from their originating python source file and line number. The errors from the compiler will however still be C++ errors but having the original source file and line gives a much better indication of what the problem is likely to be. E.g. If an error is added to an existing example on line 85 by addding `no = get_thing()` then the following error can now be produced.

```
Failed to load program for agent function (condition) 'outputdata', log:
Compilation failed: NVRTC_ERROR_COMPILATION
boids_spatial3D.py(85): error: identifier "get_thing" is undefined
      auto no = get_thing();
```

This is achieved through the use of the `#line` directive (see [more](https://learn.microsoft.com/en-us/cpp/preprocessor/hash-line-directive-c-cpp?view=msvc-170)). Line directives are inserted anywhere that the ast uses `fill` (which is the single place a new source line is created in the C++). Use of the line directive output can be optionally turned off using `bypass_line_directive` argument of `codegen.codegen()` function. This is useful is preserving the existing codegen testing suite. 

In order for the "correct" line numbers to be preserved within the code generator some hackery is required;
 - For agent functions passed as string there is no issue as line number in the ast will refer to the line number in the originating string. There is no orginating source file in this case so "PythonString" is used instead.
 - For agent functions passed as a python `callable` we previously did some funky stuff to append any agent device functions. This would obscure originating line numbers so an approach is taken to preserve these. The `ModuleExtractor` class in codegen.py will take a source module and replicate its structure creating empty lines for anything except a named agent functions and and device agent functions. This removes the requirement for us to append these as part of the codegen `translate` function. When passing a callabel it is normal for there to be a source file which is used in the line directive. There are some occasions where this is not possible (i.e. A callable form a notebook cell). In these cases a filename of "DynamicPython" is used.

The following tests have been added or updated;

 - [x] test_codegen_integration.py::test_gpu_codegen_line_error - Checks for a warning and error from a agent function calling a non exixotant function (resulting in compilation error). The error string is checked to ensure the correct filename "PythonString" and line number are reported.
 - [x] test_codegen_integration.py::test_gpu_agent_func_in_class - No longer produces a CodeGen error as translate will remove the class function (as not a global). This will produce a runtime error due to a missing FLAMEGPU_AGENT_FUNCTION instead.
 - [x] all tests in test_codegen.py - Now use `bypass_line_directive=True` during codegen to preserve compact output for cross validation.
 
A change has also been made to pyflamegpu `setup.py.in` to use an SPDX compatible licence expression rather than a license classifier which is depreciated in recent setuptools. This removes an existing warning.

### CODE REVIEW

**Review by Robadob — COMMENTED** [review]
Great change in principle with a relatively simple solution. I'm no way near familiar enough with `codegen.py` to give this a thorough review, so will leave it for @ptheywood to test on  Linux. Wouldn't surprise me if there's an obscure edge-case somewhere.

File: swig/python/setup.py.in [hunk 1]
```diff
@@ -24,11 +24,11 @@ setup(
   packages=find_namespace_packages(where='src'),
   package_dir={"": "src"},
   include_package_data=True,
+  license="AGPL-3.0-only",
```
[the remark this answers is not in the document]
Comment (mondus) [reply swig/python/setup.py.in]: Its in the PR description at the bottom. Its due to a depreciation of using classifiers to specify the license in setuptools.

File: tests/python/codegen/test_codegen_integration.py [hunk 2]
```diff
@@ -117,3 +131,25 @@ def test_gpu_codegen_function_condition(self):
             else:
                 # Not meeting the condition means the agent function will now have executed and the x value should be unchanged
                 assert x == i
+
+    def test_gpu_codegen_line_error(self):
```
Comment (ptheywood) [inline tests/python/codegen/test_codegen_integration.py]: Test suite additions only cover one of the 3 possible line based error types

- Doesn't check that a `filename.py(LINE)` error is produced
- Dosen't check that a `DynamicPython(LINE)` error is produced

**Review by ptheywood — CHANGES_REQUESTED** [review]
One typo + some nitpicking suggestions.

Test suite additions are a little light (but that matches the existing python testing, so probably not a blocker but thought it's worth raising). 

Full python test suite passed under linux with python 3.13 and CUDA 13.0 under linux. 

```
677 passed, 12 skipped in 280.69s (0:04:40)
```

Should make sure we test with 3.10/3.14 before the next release though.

File: swig/python/codegen/__init__.py [hunk 3]
```diff
@@ -1,45 +1,42 @@
...
-        # filter function by device function (modules functions are a tuple of (name: str, func: Callable))
-        d_functions = [x[1] for x in module_functions if hasattr(x[1], '__is_pyflamegpu_device_function')]
-        # get source for each function
-        for d_f in d_functions:
-            prepend_source += inspect.getsource(d_f);
-        # get source for function and prepend device functions
-        function_source = prepend_source + inspect.getsource(function)
-        tree = ast.parse(function_source)
+        # If no file attribute then its probably from a notebookc ell or some other dynamic (non file) location
+        source = "DynamicPython"
```
[the remark this answers is not in the document]
Comment (mondus) [reply swig/python/codegen/__init__.py]: Added to docs in https://github.com/FLAMEGPU/FLAMEGPU2-docs/pull/191

**Comment by mondus** [comment]
All review comments addressed and docs updated to reflect new tests and approach fro DynamicPython sources of agent functions.
